File: chapter_2_linked_lists/js/2.6-palindrome.js
const { SinglyLinkedList } = require("./singlyLinkedList")

class SinglyLinkedList_2_6 extends SinglyLinkedList {
  /* Determines if a linked list is a palindrome
   * O(n) time and space complexity */
  isPalindrome() {
    let hare = this.head
    let tortoise = this.head
    let stack = []
    /* navigate to middle node */
    while (hare && hare.next) {
      hare = hare.next.next
      stack.push(tortoise.value)
      tortoise = tortoise.next
    }
    /* for odd-lengthedlists */
    if (this.length & 1) tortoise = tortoise.next
    while (tortoise) {
      if (tortoise.value != stack.pop()) return false
      tortoise = tortoise.next
    }
    return true
  }
}

const sll = new SinglyLinkedList_2_6()
const arr = [1, 2, 3, 4, 3, 2, 1]
for (let i of arr) sll.addNode(i)
sll.printList() // 1 -> 2 -> 3 -> 4 -> 3 -> 2 -> 1
console.log(sll.isPalindrome()) // true

const sll2 = new SinglyLinkedList_2_6()
arr = [1, 2, 3, 3, 2, 1]
for (let i of arr) sll2.addNode(i)
sll2.printList() // 1 -> 2 -> 3 -> 3 -> 2 -> 1
console.log(sll2.isPalindrome()) // true

const sll3 = new SinglyLinkedList_2_6()
arr = [1, 2, 4, 3, 2, 1]
for (let i of arr) sll3.addNode(i)
sll3.printList() // 1 -> 2 -> 3 -> 4 -> 2 -> 1
console.log(sll3.isPalindrome()) // true

File: chapter_2_linked_lists/cpp/2.6-palindrome.cpp
#include "singly_linked_lists.h"
#include <vector>

class LinkedList2_6 : public LinkedList
{
public:
  bool is_palindrome();
};

/** 
 * Determines if a linked list is a palindrome
 * O(n) time and space complexity
 */
bool LinkedList2_6::is_palindrome()
{
  Node *curr;
  int mid;
  vector<int> stack;

  mid = this->length / 2;
  curr = this->head;
  while (mid)
  {
    stack.push_back(curr->value);
    curr = curr->next;
    mid--;
  }
  if (this->length & 1)
    /* move pointer one node if list length is odd */
    curr = curr->next;
  while (curr)
  {
    if (curr->value != stack.back())
      return false;
    stack.pop_back();
    curr = curr->next;
  }
  return true;
}

int main()
{
  LinkedList2_6 *ll = new LinkedList2_6();
  int keys[] = {1, 2, 4, 5, 4, 2, 1};
  int len = sizeof(keys) / sizeof(int);
  for (int i = 0; i < len; i++)
    ll->add_node(keys[i]);
  ll->print_list();                    // 1 -> 2 -> 4 -> 5 -> 4 -> 2 -> 1
  cout << ll->is_palindrome() << endl; // 1

  LinkedList2_6 *ll2 = new LinkedList2_6();
  int keys2[] = {1, 2, 4, 4, 2, 1};
  len = sizeof(keys2) / sizeof(int);
  for (int i = 0; i < len; i++)
    ll2->add_node(keys2[i]);
  ll2->print_list();                    // 1 -> 2 -> 4 -> 4 -> 2 -> 1
  cout << ll2->is_palindrome() << endl; // 1

  LinkedList2_6 *ll3 = new LinkedList2_6();
  int keys3[] = {1, 2, 4, 5, 6, 4, 2, 1};
  len = sizeof(keys3) / sizeof(int);
  for (int i = 0; i < len; i++)
    ll3->add_node(keys3[i]);
  ll3->print_list();                    // 1 -> 2 -> 4 -> 6 -> 5 -> 4 -> 2 -> 1
  cout << ll3->is_palindrome() << endl; // 0

  return 0;
}